libcontainer/cgroups/v1_utils.go

```diff
@@ -91,6 +95,16 @@ func tryDefaultPath(cgroupPath, subsystem string) string {
 	return path
 }
 
+func ReadCgroupMountinfo() ([]*mountinfo.Info, error) {
```

Comment:
```suggestion
// ReadCgroupMountinfo returns a list of cgroup mounts for the current running
// process. ReadCgroupMountinfo caches the results in memory, and assumes that
// no new cgroup mounts appear after it was first called.
func ReadCgroupMountinfo() ([]*mountinfo.Info, error) {
```

(perhaps "a list of cgroup v1 mounts"?)

Reply:
Actually; the name `ReadCgroupMountinfo` is now a bit ambiguous; `Read....` would indicate it actually _reads_ (not caches) `mountinfo`.

Looking a bit further; instead of introducing a new, exported, function;

- This same file has  a `getCgroupMountsV1()`, which is called by `GetCgroupMounts()`
- `getCgroupMountsV1()` performs additional conversion (mapping `mountinfo.Info` to a `libcontainer/cgroups.Mount` using `getCgroupMountsHelper()`)
    - this conversion will be done _every time_ `GetCgroupMounts()` is called
- *side-note*: I see `` and `getCgroupMountsV1()` have an `all` argument, that seems to be always `false` (perhaps it's no longer needed)
- All current uses of `ReadCgroupMountinfo()` outside of this package (`libcontainer/cgroups/fs/cpuset.go` and `libcontainer/cgroups/fs/fs.go` are only interested in `Mountpoint` and `Root`, _both_ of which are also preserved in `libcontainer/cgroups.Mount` (returned by `getCgroupMountsV1 ()`


Perhaps:

- un-export `ReadCgroupMountinfo()`
- remove the `sync.Once()` from `ReadCgroupMountinfo()`
- _export_ `getCgroupMountsV1` (perhaps rename to `GetCgroupV1Mounts()`
    - This makes it a nice parallel to `GetCgroupMounts()` (but only returning `v1` mounts, and thus clearer on what it does)
- remove the `all` argument (if it's indeed no longer needed)
- add the `sync.Once` to `GetCgroupV1Mounts()`, so that the conversions are only done once (if no new cgroups are expected to be added, there should be no need to perform the conversion each time either?)

This also makes the interface somewhat cleaner, as we're not returning an "external" type (`mountinfo.Info`, but instead a type that's defined in this package (`cgroups.Mount`)

Reply:
1. `all` was introduced by https://github.com/opencontainers/runc/pull/1049, and it very vividly demonstrates the major problem with runc: it's a tool as well as a set of packages with lots of users. I have changed runc/libcontainer public API in the past and in some cases the users had to modify their code, and in some cases I needed to revert back.

I will look into cadvisor code wrt using `all` argument.

2. The idea of reusing GetCgroupMounts (and adding sync.Once to it) looks interesting. The problem is, it's somewhat harder to do with `all` argument as it is now. I tried it in the past, and the result was lots of code. What I like about this PR is it is simple, it removes lots of code making things faster at the same time.

3. I don't like yet another exported function either. I think the solution to that is to use `internal`.

Reply:
> I will look into cadvisor code wrt using all argument.

https://github.com/google/cadvisor/issues/1461
https://github.com/google/cadvisor/pull/1476

I read all that and thought for some time and I am still not sure how to handle this :(

Reply:
So, I have added a comment describing `ReadCgroupMountinfo` as per your earlier suggestion, and I'll take a look at internalizing it later.

Reply:
I agree with doing this API cleanup later.

Reply:
OK, I took a fresh look and found a way to unexport `ReadCgroupMountinfo` by not using in the other two commits. The core of this PR stays the same, but the last two commits (getCgroupRoot and cpusets ones) changed a lot. PR description updated with "v2 changes", please re-review @thaJeztah @cyphar